src/common/libsdprocess/sdprocess.c

```diff
@@ -0,0 +1,1835 @@
+                                           int stderr_fd)
+{
+    struct flux_sdprocess *sdp = NULL;
+    /* note that lower case should be used, not upper case hex.  Apparently
+     * a part of dbus standard.
+     */
+    char hex2char[16] = "0123456789abcdef";
+    const char *service_path_prefix = "/org/freedesktop/systemd1/unit/";
+    int service_path_prefix_len = strlen (service_path_prefix);
+    /* _2e is escape of . (period) */
+    const char *service_path_suffix = "_2eservice";
+    int service_path_suffix_len = strlen (service_path_suffix);
+    const char *ptr;
+    int index, len, ret;
+
+    if (!(sdp = calloc (1, sizeof (*sdp))))
+        return NULL;
+    if (!(sdp->unitname = strdup (unitname)))
+        goto cleanup;
+    if (strv_copy (argv, &sdp->argv) < 0)
+        goto cleanup;
+    if (strv_copy (envv, &sdp->envv) < 0)
+        goto cleanup;
+    sdp->h = h;
+    sdp->reactor = reactor;
+    sdp->stdin_fd = stdin_fd;
+    sdp->stdout_fd = stdout_fd;
+    sdp->stderr_fd = stderr_fd;
+
+    if ((ret = sd_bus_open_user (&sdp->bus)) < 0) {
+        errno = -ret;
+        goto cleanup;
+    }
+    if (asprintf (&sdp->service_name, "%s.service", sdp->unitname) < 0)
+        goto cleanup;
+    /* we must escape all special chars b/c dbus cannot handle them.
+     * escape rule is <special char> -> _<hex bytes of char>.
+     * e.g. . -> _2e
+     *
+     * note that a-f should be lower case
+     *
+     * also escape 0-9 if numeral is first character in unitname.
+     */
+    len = service_path_prefix_len;
+    len += strlen (sdp->unitname) * 3;
+    len += service_path_suffix_len;
+    len += 1;
+    if (!(sdp->service_path = calloc (1, len)))
+        goto cleanup;
+    strcpy (sdp->service_path, service_path_prefix);
+    index = service_path_prefix_len;
+    ptr = sdp->unitname;
+    while (*ptr) {
+        if (!((*ptr) >= 'A' && (*ptr) <= 'Z')
+            && !((*ptr) >= 'a' && (*ptr) <= 'z')
+            && !(ptr > sdp->unitname && (*ptr) >= '0' && (*ptr) <= '9')) {
+            sdp->service_path[index++] = '_';
+            sdp->service_path[index++] = hex2char[*ptr >> 4];
+            sdp->service_path[index++] = hex2char[*ptr & 0xF];
+        }
+        else
+            sdp->service_path[index++] = *ptr;
+        ptr++;
+    }
+    strcpy (sdp->service_path + index, service_path_suffix);
+
+    return sdp;
+
+cleanup:
+    flux_sdprocess_destroy (sdp);
+    return NULL;
+}
+
+static int transient_service_set_properties (flux_sdprocess_t *sdp,
+                                             sd_bus_message *m)
+{
+    int ret;
+
+    if ((ret = sd_bus_message_append (m,
+                                      "(sv)",
+                                      "Description",
+                                      "s",
+                                      "libsdprocess")) < 0) {
+        set_and_log_errno (sdp->h, ret, "sd_bus_message_append");
+        return -1;
+    }
```

Comment:
There is so much repetition in this function....does each error need to be separately logged or could they be consolidated with a goto that doesn't call out the function that failed?